File: Java/Hashing/Hashing.java
import java.util.*;

public class Hashing {
    public static void countDigits(int arr[], int n) {
        var map = new HashMap<Integer, Integer>();
        for (int i = 0; i < n; i++)
            map.put(arr[i], map.getOrDefault(arr[i], 0) + 1);

        for (int i : map.keySet())
            System.out.println("number: " + i + " is repeated " + map.get(i) + " times.");
    }

    public static void countFreq(int arr[], int n) {
        boolean visited[] = new boolean[n];
        int maxFreq = 0, minFreq = n;
        int maxEle = 0, minEle = 0;
        for (int i = 0; i < n; i++) {
            // Skip this element if already possessed.
            if (visited[i] == true) {
                continue;
            }
            ;

            // Count frequency
            int count = 1;
            for (int j = i + 1; j < n; j++) {
                if (arr[i] == arr[j]) {
                    visited[j] = true;
                    count++;
                }
            }
            if (maxFreq < count) {
                maxEle = arr[i];
                maxFreq = count;
            }
            if (count < minFreq) {
                minEle = arr[i];
                minFreq = count;
            }
        }
        System.out.println("The highest frequency element is: " + maxEle);
        System.out.println("The lowest frequency element is: " + minEle);
    }

    public static void main(String args[]) {
        int arr[] = { 1, 1, 1, 3, 6, 7, 6, 8, 7 };
        countFreq(arr, arr.length);

    }
}

File: cpp/Hashing/hashing.cpp
#include <iostream>
#include <vector>
using namespace std;

void countFreq(int arr[], int n)
{
    vector<bool> visited(n, false);
    for (int i = 0; i < n; i++)
    {
        // skip this element if already processed
        if (visited[i] == true)
            continue;

        // count frequency
        int count = 1;
        for (int j = i + 1; j < n; j++)
        {
            if (arr[i] == arr[j])
            {
                visited[j] = true;
                count++;
            }
        }
        cout << arr[i] << " " << count << endl;
    }
}

void countFreq2(int arr[], int n)
{
    vector<bool> visited(n, false);
    for (int i = 0; i < n; i++)
    {
        if (visited[i] == true)
            continue;

        int count = 1;
        for (int j = i + 1; j < n; j++)
        {
            if (arr[i] == arr[j])
            {
                visited[j] = true;
                count++;
            }
        }

        cout << arr[i] << " " << count << endl;
    }
};

void countFreq3(int arr[], int n)
{
    unordered_map<int, int> map;
    for (int i = 0; i < n; i++)
        map[arr[i]]++;

    // Traverse through map and print frequencies
    for (auto x : map)
    {
        cout << x.first << " -- " << x.second << endl;
    }
}

void maxFreq(vector<int> &nums, int n)
{
    vector<int> flags(n, false);
    int maxFreq = 0, maxEl = 0;
    int minEl = 0, minFreq = n;

    for (int i = 0; i < n; i++)
    {
        if (flags[i] == true)
            continue;

        int count = 1;
        for (int j = i + 1; j < n; j++)
        {
            if (nums[i] == nums[j])
            {
                flags[j] = true;
                count++;
            }
        }

        if (maxFreq < count)
        {
            maxEl = nums[i];
            maxFreq = count;
        }
        if (minFreq > count)
        {
            minFreq = count;
            minEl = nums[i];
        }
    }

    cout << "the max element is " << maxEl << endl;
    cout << "the min element is " << minEl << endl;
}

void Frequency(vector<int> arr, int n)
{
    unordered_map<int, int> map;

    for (int i = 0; i < n; i++)
    {
        map[arr[i]]++;
    }
    int maxFreq = 0, maxEl = 0;
    int minFreq = n, minEl = 0;

    for (auto it : map)
    {
        int count = it.second;
        int element = it.first;

        if (count > maxFreq)
        {
            maxEl = element;
            maxFreq = count;
        }

        if (count < minFreq)
        {
            minFreq = count;
            minEl = element;
        }
    }

    cout << "The highest element is : " << maxEl << endl;
    cout << "The lowest element is : " << minEl << endl;
}

int main()
{
    vector<int> arr = {10, 5, 10, 15, 10, 5};
    int n = sizeof(arr) / sizeof(arr[0]);
    // countFreq2(arr, n);
    cout << "---------- " << endl;
    Frequency(arr, n);

    unordered_map<int, int> map;

    for (auto i = 0; i < arr.size(); i++)
    {
        map[arr[i]]++;
    }

    for (auto it : map)
    {
        cout << it.first << " " << it.second << endl;
    }
    return 0;
}